src/index.d.ts

```diff
@@ -1,966 +1,54 @@
-export as namespace Reselect;
-
-export type Selector<S, R> = (state: S) => R;
-
-export type OutputSelector<S, R, C> = Selector<S, R> & {
-  resultFunc: C;
-  recomputations: () => number;
-  resetRecomputations: () => number;
-}
+// Minimum TypeScript Version: 4.2
 
-export type ParametricSelector<S, P, R> = (state: S, props: P, ...args: any[]) => R;
+export as namespace Reselect;
 
-export type OutputParametricSelector<S, P, R, C> = ParametricSelector<S, P, R> & {
-  resultFunc: C;
+export type Selector<S = any, R = unknown, P extends never | (readonly any[]) = any[]> = [P] extends [never] ? (state: S) => R : (state: S, ...params: P) => R;
+export type OutputSelector<S extends SelectorArray, Result, Params extends readonly any[], Combiner> = Selector<GetStateFromSelectors<S>, Result, Params> & {
```

Comment:
In the idea of backwards compatibility, make have `Combiner` stay the third argument and move `Params` to the fourth?

Reply:
Yes! Good catch. I'm not even sure why I made that change.

Reply:
I'll have this fixed in the next push.